Add T_CRITICAL_SECTION for cross OS DAC compile

sdmaclea opened the pull request:
Pad Crst union size to keep layout same when cross OS compiling DAC

@jkotas This was my original implementation of the fix required to get `class Crsrt` consistent for the cross OS DAC builds.

I know we talked about moving it into crosscompile.h, but it is pretty simple here...  The BOTR recommend avoiding direct use of CRITICAL_SECTION, so this is the central use point.

I'll start drafting the crosscompile.h version tomorrow, unless you think this is PR acceptable.

sdmaclea committed:
T_CRITICAL_SECTION for DAC

Add T_CRITICAL_SECTION as a cross OS version of CRITICAL_SECTION
Add platform defines for TARGET_<os> as needed.
Add verification code to keep T_CRITICAL_SECTION in sync with CRITICAL_SECTION

jkotas commented on src/coreclr/src/vm/crst.h:
```diff
@@ -298,6 +298,15 @@ friend class Crst;
     union {
         CRITICAL_SECTION    m_criticalsection;
+#if defined(TARGET_UNIX)
+	// m_criticalsection is exactly 0x90 w/ TARGET_UNIX, but smaller on Windows.
```
I do not think that this is right: https://github.com/dotnet/runtime/blob/f44c7d1e8ddc48ab348871c25b65ca031aa0c3ba/src/coreclr/src/pal/inc/pal.h#L2348

sdmaclea replied:
I just fixed the `#ifdef` last night.  I was kind of surprised the data structure was consistent across arch and build type....  It was simply because the code was being ignored.

Thanks for pointing out the PAL structure.  Having it make it easier to place a Windows version  in crosscompile.h

jkotas commented:
> I'll start drafting the crosscompile.h version tomorrow, unless you think this is PR acceptable.

If it is just a tiny bit like in the current delta, I think it would be ok ... but I am not sure how much worse it is going to get once it actually compiles and works.

jkotas commented on src/coreclr/src/inc/crosscomp.h:
```diff
@@ -393,6 +393,60 @@ typedef struct _T_KNONVOLATILE_CONTEXT_POINTERS {
+    LONG LockCount;
+    LONG RecursionCount;
+    HANDLE OwningThread;
+    HANDLE LockSemaphore;
```
`LockSemaphore` and `bInternal` fields do not seem to be used for anything. You can delete them while you are on it.

jkotas replied:
`dwInitState` as well.

sdmaclea replied:
Looks like CRITICAL_SECTION is also shadowed by PAL_CRITICAL_SECTION in src/pal/src/sync/cs.cpp.  The disabled debug `PAL_TRACK_CRITICAL_SECTIONS_DATA` code is using `bInternal` & `dwInitState`.  

I will leave those members, but put them under the same `#if`

sdmaclea replied:
I must have gotten distracted checking *InitState, because it is used in asserts all the time...

sdmaclea committed:
Remove unused CRITICAL_SECTION fields

sdmaclea committed:
Revise PAL_CRITICAL_SECTION

Remove LockSemaphore
Restore CRITICAL_SECTION/T_CRITICAL_SECTION debug members under #if

jkotas commented on src/coreclr/src/pal/inc/pal.h:
```diff
@@ -2379,6 +2379,11 @@ typedef struct _CRITICAL_SECTION {
+#ifdef PAL_TRACK_CRITICAL_SECTIONS_DATA
+    BOOL bInternal;
+    volatile DWORD dwInitState;
+#endif PAL_TRACK_CRITICAL_SECTIONS_DATA
```
```suggestion
#endif // PAL_TRACK_CRITICAL_SECTIONS_DATA
```

sdmaclea committed:
Oops

sdmaclea committed:
Restore InitState

sdmaclea committed:
One more fix

am11 commented on src/coreclr/src/inc/crosscomp.h:
```diff
@@ -393,6 +393,62 @@ typedef struct _T_KNONVOLATILE_CONTEXT_POINTERS {
+#define DAC_CS_NATIVE_DATA_SIZE 56
+#else
+#warning
+#error  DAC_CS_NATIVE_DATA_SIZE is not defined for this architecture
```
@sdmaclea, could you please specify which struct this size correspond to?
I have ported `PAL_CS_NATIVE_DATA_SIZE` in `pal.h` to Solaris and it was clear that we are dealing with size of `struct PAL_CS_NATIVE_DATA` (which in case of solaris x64 is 48), but here I couldn't find `struct DAC_CS_NATIVE_DATA`. If it is the size of `struct T_CRITICAL_SECTION`, should we consider renaming it to `T_CRITICAL_SECTION_SIZE` to make it self-explanatory?

sdmaclea replied:
See https://github.com/dotnet/runtime/search?q=PAL_CS_NATIVE_DATA_SIZE&unscoped_q=PAL_CS_NATIVE_DATA_SIZE

It is the DAC's copy of `PAL_CS_NATIVE_DATA_SIZE`.  You can see it in the assert in src/coreclr/src/vm/crst.cpp. They are supposed to be the same value.

If you want to add a comment to that effect, it seems like a good idea. I assume this could easily go in your solaris PR. If you would rather I do it, I can create a PR.

am11 replied:
Thank you for the information. I will add it to my Solaris port branch. I was thinking to update the error message to the effect “what to look for”, as this error case is to mainly assist folks porting code to new platforms. I still have some code base to cover, so it may take few days for me to prepare a PR.
